tmt/utils.py

```diff
@@ -1006,8 +1006,93 @@ def environment_to_dict(variables: Union[str, List[str]]) -> EnvironmentType:
     return result
 
 
-def environment_file_to_dict(
-        env_files: Iterable[str], root: str = ".") -> Dict[str, str]:
+@lru_cache(maxsize=None)
+def environment_file_to_dict(env_file: str, root: str = ".") -> EnvironmentType:
+    """
+    Create dict from a file.
+
+    File should be in yaml/yml or dotenv format.
+
+    dotenv file example:
+        ```bash
+        A=B
+        C=D
+        ```
+
+    yaml file example:
+        ```yaml
+        A: B
+        C: D
+        ```
+
+    Path to the file should be relative to the metadata tree root.
+    """
+
+    env_file = env_file.strip()
+
+    # Fetch a remote file
+    if env_file.startswith("http"):
+        # Create retry session for longer retries, see #1229
+        session = retry_session.create(
+            retries=ENVFILE_RETRY_SESSION_RETRIES,
+            backoff_factor=ENVFILE_RETRY_SESSION_BACKOFF_FACTOR,
+            allowed_methods=('GET',),
+            status_forcelist=(
+                429,  # Too Many Requests
+                500,  # Internal Server Error
+                502,  # Bad Gateway
+                503,  # Service Unavailable
+                504   # Gateway Timeout
+                ),
+            )
+        try:
+            response = session.get(env_file)
+            response.raise_for_status()
+            content = response.text
+        except requests.RequestException as error:
+            raise GeneralError(
+                f"Failed to fetch the environment file from '{env_file}'. "
+                f"The problem was: '{error}'")
+
+    # Read a local file
+    else:
+        # Ensure we don't escape from the metadata tree root
+        try:
+            root_path = Path(root).resolve()
+            full_path = (Path(root_path) / Path(env_file)).resolve()
+            full_path.relative_to(root_path)
+        except ValueError:
+            raise GeneralError(
+                f"The 'environment-file' path '{full_path}' is outside "
+                f"of the metadata tree root '{root}'.")
+        if not Path(full_path).is_file():
+            raise GeneralError(f"File '{full_path}' doesn't exist.")
+
+        content = Path(full_path).read_text()
+
+    # Parse yaml file
+    if os.path.splitext(env_file)[1].lower() in ('.yaml', '.yml'):
+        environment = parse_yaml(content)
+
+    else:
+        try:
+            environment = parse_dotenv(content)
+
+        except ValueError:
+            raise GeneralError(
+                f"Failed to extract variables from environment file "
+                f"'{full_path}'. Ensure it has the proper format "
+                f"(i.e. A=B).")
+
+    if not environment:
+        log.warn(f"Empty environment file '{env_file}'.")
+
+        return {}
+
+    return environment
+
+
+def environment_files_to_dict(env_files: Iterable[str], root: str = ".") -> Dict[str, str]:
     """
     Create dict from files.
```

Comment:
Could docstring point to environment_file_to_dict and just mention that this combines several files together?
Most of the docsting is the same and IMO without additional value

Reply:
I tried to improve in [344d5f9](https://github.com/teemtee/tmt/pull/1313/commits/344d5f9df12dbf364c45a7fe64ac9ae95439ada1), let me know if it's better. There will be an overlap, but I tried to add notes and explain the difference.